### DISCUSSION TITLE
General ideas

### DISCUSSION THREAD

**Discussion by thorwhalen** [thorwhalen/allude#3] [Ideas]
The goal is to get from ideation to implementation fast. Cut through the boilerplate, and enable quick feedback to facilitate a conversation between both. When implementation (any generation/creation of something representing the idea) takes to long to come to life, the ideation process looses its' momentum and its' perspective. This is reminiscent of the agile framework, and tapping into it might be helpful for some angles -- but it's important to start afresh too, because much agility isn't agile anymore: it has been codified and/or institutionalized.

## My work

A lot of my efforts in architecture have been oriented in making it fast to roll out implementations fast, but in a way that is more "additive" and less "entropic". Without even knowing it most of the time, I gravitated to principles that had already been identified, named, and milked by the masters of software. This is encouraging, and quite fortunate, since my focus was on speed and flexibility. I even (horrifying some around me) had the attitude of "I don't care if it's fragile, as long as I get it now -- I'll address fragilities if and when needed". It seemed like my desire for agility would be in conflict with the need for robustness and maintainability moving forward. Yet, I've discovered, the same principles underpinned both. 

Essentially, anytime I had to do something (or my developers had to do something), and that something took longer than I thought it should, I would ask myself the questions:
* How often might I want to do that thing? And if the answer was "often enough"...
* What boilerplate can be eliminated? What can be automated? What is the simplest way to think about this (moving the devil to the details)?

This has brought me to work on several fronts, that I've identified are common enough to warrant making it easy and agile:
* Data operations. Making it easy to
  * Find data
  * Access data
  * Prepare data (shape it for the domain/business_logic)
  * Change data backend
  * Store stuff (without having to think of data architecture first)
* Functionality
  * Compose objects (classes, functions, etc.) together
  * Separate execution model (declarative form + execution model)
  * Facilitate functionality CRUD (add stuff, replace stuff, remove stuff, group stuff,...)
* Create User Interfaces. I got my functionality -- now where is my "product"?
  * Make CLIs
  * Make webservices
  * Make GUIs

## The new era

But now, there's AI.

I always thought that more time should be spent focusing AI development on AI development itself. 
That is, figure out how AI can help us in the tasks that we undertake when developing AI tools. 
Things like implementing ideas in code, trying ideas out, etc.

I [jotted ideas about that down](https://github.com/thorwhalen/essays/blob/master/ai_horizons.md#ideas-to-consider), and tried to discuss it with people, but didn't get the most encouraging feedback. The AI community thought I was being far fetched, and the developer community thought I was just being insane! Self-organizing code!? Don't get near my repositories!

Whether it was that, or the fact that I needed to be a responsible and pragmatic manager, I ended up thinking a lot more about how to achieve what I wanted through design and good engineering. I don't resent it: It taught me a lot. 

But now... there's AI!

It didn't come from where I expected (I'm not a "black box" deep learning fan), but boy did it come. 

And now, we really need to rethink things. What will software development look like in the near future?

**Comment by jklafdjald** [comment]
My first project in my life. I want to grow my this website [https://snowdaypredictorcanada.com](https://snowdaypredictorcanada.com)
